src/puppetlabs/ezbake/core.clj

```diff
@@ -501,23 +541,112 @@ Dependency tree:
 
 (defn generate-project-data-yaml
-  [lein-project build-target]
+  [lein-project build-target additional-uberjars]
   (lein-main/info "generating project_data.yaml file")
   (spit
     (fs/file staging-dir "ext" "project_data.yaml")
     (stencil/render-string
       (slurp (get-staging-template-file "project_data.yaml.mustache"))
-      {:project       (:name lein-project)
-       :summary       (:description lein-project)
-       :description   (format "%s (%s)"
-                              (:description lein-project)
-                              (deputils/generate-manifest-string lein-project))
-       :uberjar-name  (:uberjar-name lein-project)
-       :is-pe-build   (format "%s" (= (get-local-ezbake-var lein-project :build-type "foss") "pe"))
-       :repo-name     (format "%s" (get-local-ezbake-var lein-project :repo-target ""))})))
+      {:project (:name lein-project)
+       :summary (:description lein-project)
+       :description (format "%s (%s)"
+                            (:description lein-project)
+                            (deputils/generate-manifest-string lein-project))
+       :uberjar-name (:uberjar-name lein-project)
+       :additional-uberjars (mapv (fn [filename] {:uberjar filename}) additional-uberjars)
+       :is-pe-build (format "%s" (= (get-local-ezbake-var lein-project :build-type "foss") "pe"))
+       :repo-name (format "%s" (get-local-ezbake-var lein-project :repo-target ""))})))
+
+(schema/defn get-additional-uberjars
+  "Returns the list of additional uberjar dependencies from the given lein project"
+  [lein-project]
+  (when-let [dependencies-vector (get-in lein-project [:lein-ezbake :additional-uberjars])]
+    dependencies-vector))
+
+(schema/defn resolve-dependency! :- File
+  "Resolves a single dependency and returns a File pointing to it's jar.
+
+  This has the side effect of fetching the whole dependency tree for the given
+  dependency, but we only care about the one jar"
+  [[project-symbol version :as dependency-coordinates]
+   repositories]
+  (lein-main/info "Resolving dependency for " dependency-coordinates)
+  (let [project {:dependencies [dependency-coordinates]
+                 :repositories repositories}
+        ; resolved-dependencies will be a list of jars, one for each dependency
+        ; in the project, including the jar for this project
+        resolved-dependencies (lein-classpath/resolve-managed-dependencies :dependencies nil project)
+        jar-regex (re-pattern (format "%s-%s.jar" (name project-symbol) version))
+        jar-file (first (filter #(re-find jar-regex (.getName %))
+                                resolved-dependencies))]
+    jar-file))
+
+(schema/defn extract-project-from-jar! :- schema/Str
+  "Extracts the project.clj file out of the given jar into the destination directory"
+  [jar :- File
+   destination-dir :- schema/Str]
+  (fs/mkdirs destination-dir)
+  (let [jar-file (JarFile. jar)
+        jar-entry (.getJarEntry jar-file "project.clj")
```

Comment:
The current work for the `additional-uberjars` is designed around retrieving repo artifacts and building an uberjar from them.  The jars deployed to repos for an artifact, including how we're handling the jruby-deps ones for Puppet Server, are typically not uberjars themselves.  They are just jars with compiled elements from their immediate project but with additional metadata indicating their upstream dependencies.  What we want to package into ezbake, though, are uberjars built from the immediate project jars which have been fetched from the repo.

For Clojure-based projects, building the uberjar from the project.clj is helpful because it can include any special packaging requirements, e.g., the uberjar-exclusions that we want for JRuby uberjars for Puppet Server.  It probably would be possible to extend this to build uberjars for pure-Java projects, e.g., using the pom file from the jar.  Not sure if this could be done with leiningen, though, or if we'd have to use Maven instead?  I definitely didn't want to try to get into that at this point until we had a specific use case for it.

Another reasonable extension to all of this at some point may be a way to, instead specifying coordinates for repo artifacts to build the uberjar from, provide a fully-qualified URL to some location where pre-built uberjars may have been distributed.  That could be pretty general-purpose (non-Clojure/Java specific).  That might also be nice for the JRuby uberjars at some point as well but I didn't want to force us to work through all of the delivery logistics for that at this point.